d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,709 @@
+  private final Map<String, ResourceSubscriber> _cdsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> _edsResourceSubscribers = new HashMap<>();
+
+  private AdsStream _adsStream;
+
+  public XdsClientImpl(Node node, ManagedChannel managedChannel, ExecutorService executorService)
+  {
+    _node = node;
+    _managedChannel = managedChannel;
+    _executorService = executorService;
+  }
+
+  private void startRpcStream()
+  {
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(_managedChannel);
+    _adsStream = new AdsStream(stub);
+    _adsStream.start();
+    _log.info("ADS stream started");
+  }
+
+  @Override
+  void watchLdsResource(final String resourceName, final LdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _ldsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe LDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.LDS, resourceName);
+        _ldsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.LDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchRdsResource(final String resourceName, final RdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _rdsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe RDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.RDS, resourceName);
+        _rdsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.RDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchCdsResource(final String resourceName, final CdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _cdsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe CDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.CDS, resourceName);
+        _cdsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.CDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchEdsResource(final String resourceName, final EdsResourceWatcher watcher)
```

Comment:
The watch*dsResouce() methods seem to be repetitive, can we try parameterizing & consolidating them?

Reply:
Good point. Consolidated